File: core/survey.go
package core

import (
	"fmt"

	"github.com/AlecAivazis/survey/v2"
)

func SurveyReleases(repo *Repo, releases []Release) ([]Asset, error) {
	var releaseTitles []string
	var releaseTitleMap = make(map[string]Release)
	for _, release := range releases {
		title := release.Title()
		releaseTitles = append(releaseTitles, title)
		releaseTitleMap[title] = release
	}

	// ask select releases
	var q1 = &survey.Select{
		Message: fmt.Sprintf("Select releases of %s:", repo.String()),
		Options: releaseTitles,
	}
	var selectedReleaseName string
	err := survey.AskOne(q1, &selectedReleaseName)
	if err != nil {
		return nil, err
	}

	if selectedReleaseName == "" {
		return nil, fmt.Errorf("nothing selected")
	}
	selectedRelease := releaseTitleMap[selectedReleaseName]

	var assetTitles []string
	var assetTitleMap = make(map[string]Asset)
	for _, asset := range selectedRelease.Assets {
		title := asset.Title()
		assetTitles = append(assetTitles, title)
		assetTitleMap[title] = asset
	}

	// ask select assets
	var q2 = &survey.MultiSelect{
		Message: "Select assets:",
		Options: assetTitles,
	}
	var selectedAssetNames []string
	err = survey.AskOne(q2, &selectedAssetNames)
	if err != nil {
		return nil, err
	}
	if len(selectedAssetNames) == 0 {
		return nil, fmt.Errorf("nothing selected")
	}
	var selectedAssets []Asset
	for _, assetName := range selectedAssetNames {
		asset := assetTitleMap[assetName]
		selectedAssets = append(selectedAssets, asset)
	}

	return selectedAssets, nil
}

func SurveyCache(repos []string) ([]string, error) {
	// ask select repos
	var q1 = &survey.MultiSelect{
		Message: "Select repos:",
		Options: repos,
	}
	var selected []string
	err := survey.AskOne(q1, &selected)
	if err != nil {
		return nil, err
	}
	if len(selected) == 0 {
		return nil, fmt.Errorf("nothing selected")
	}
	return selected, nil
}

File: src/gh2/survey.py
from typing import Sequence, Set

from survey import routines

from ._download import Asset, Release
from .parser import Repo


def survey_releases(repo: Repo, releases: Sequence[Release]) -> Sequence[Asset]:
    # why pyright infers this as None? its real type is Set[int].
    indexes: Set[int] = routines.basket(  # type: ignore
        f"select releases of {repo}: ",
        options=map(str, releases),
    )
    selected_releases = (releases[index] for index in indexes)

    selected_assets = []
    for release in selected_releases:
        assets = release.assets
        indexes = routines.basket(  # type: ignore
            f"select asset ({len(assets)}): ",
            options=map(str, assets),
        )
        selected_assets.extend(assets[index] for index in indexes)

    return selected_assets


def survey_cache(repos: Sequence[str]) -> Sequence[str]:
    indexes: Set[int] = routines.basket(  # type: ignore
        "select repos: ",
        options=repos,
    )
    selected_repos = [repos[index] for index in indexes]

    return selected_repos
